docs/common/use-table-of-contents.ts

```diff
@@ -0,0 +1,109 @@
+import { useRouter } from 'next/router';
+import { useCallback, useEffect, useState } from 'react';
+
+type TableOfContentsOptions = {
+  /** The root containing element to search for headings and observe them scrolling */
+  root?: Element;
+  /** The query selector to fetch all headings, defaults to `h2, h3` */
+  selector?: string;
+};
+
+/**
+ * Find the headings within an element to generate a table of contents.
+ * This also uses an intersection observer to detect whenever headings are visible.
+ *
+ * @todo Remove "first flash of unstyled content" by extracting headings from MDX
+ */
+export function useTableOfContents(options: TableOfContentsOptions = {}) {
+  const [activeId, setActiveId] = useState<string>();
+  const headings = useHeadings(options.root, options.selector);
+  const router = useRouter();
+
+  useObserver(options.root, setActiveId, headings);
+
+  const onRouteChange = useCallback(
+    (url: string) => setActiveId(url.split('#', 2).pop()!),
+    [setActiveId]
+  );
+
+  useEffect(function onDidMount() {
+    router.events.on('hashChangeStart', onRouteChange);
+    if (window.location.hash) {
+      onRouteChange(window.location.hash);
```

Comment:
Probably should move this to only use `next/router` stuff, instead of `window.location`.